Modify rule S5131: Specific patch instructions for Django Template Language (#959)
SonarSource/rspec: rules/S5131/python/dtl/rule.adoc

== How to fix it?

The following code is vulnerable to Cross-Site Scripting because Jinja auto-escaping of special HTML characters has been disabled.

[source,python]
----
from django.shortcuts import render

def hello(request):
        name = request.GET.get("name")
        return render(request, 'hello.html', {'hello': hello})
----

[source,html]
----
<!doctype html>
{% autoescape false %} 
{{ hello }} <!-- Noncompliant -->
{% endautoescape %}
----

> Stop disabling auto-escaping by moving the dynamic HTML content to the template engine.

Reviewer: ```suggestion
Enable auto-escaping by moving the dynamic HTML content to the template engine.
```

> [[source,python]

Reviewer: ```suggestion
[source,python]
```

=== Pitfalls

Although auto-escaping drastically decreases the chance of introducing Cross-Site Scripting vulnerabilities, there are still specific cases where vulnerabilities can occur.

==== Unquoted variables in attributes

If ``++color++`` is untrusted, JavaScript code can be injected by adding event-handler attributes like ``++onload++``.

[source,html]
----
<!doctype html>
<span style={{color}}>...</span>
----

Surround the attribute with double or single quotes to prevent arbitrary HTML tags from being added.

[source,html]
----
<!doctype html>
<span style="{{color}}">...</span>
----

==== Variables in ``++href++`` attributes

If ``++url++`` is untrusted, JavaScript code can be injected by using an attributes value that starts with ``++javascript:++``.

[source,html]
----
<!doctype html>
<a href="{{url}}">...</a>
----

Use functions like ``++{% url %}++`` tag to generate URLs used in ``++href++`` attributes

[source,html]
----
<!doctype html>
<a href="{% url 'index' %}">...</a>
----

> ==== Variables in a script blocks

Reviewer: ```suggestion
==== Variables in script blocks
```
